models/exposure.py

```diff
@@ -1,7 +1,238 @@
+
+    mjd = sa.Column(
+        sa.Double,
+        nullable=False,
+        index=True,
+        doc="Modified Julian date of the exposure. (MJD=JD-2400000.5)"
+    )
+
+    exp_time = sa.Column(sa.Float, nullable=False, index=True, doc="Exposure time in seconds")
+
+    filter = sa.Column(sa.Text, nullable=False, index=True, doc="Filter name")
+
+    instrument = sa.Column(sa.Text, nullable=False, index=True, doc="Instrument name")
+
+    telescope = sa.Column(sa.Text, nullable=False, index=True, doc="Telescope name")
+
+    num_ccds = sa.Column(sa.Integer, nullable=False, index=False, doc="Number of CCDs / sections in the full field. ")
+
+    width = sa.Column(sa.Integer, nullable=False, index=False, doc="Width of each CCD in pixels. ")
+
+    height = sa.Column(sa.Integer, nullable=False, index=False, doc="Height of each CCD in pixels. ")
+
+    project = sa.Column(sa.Text, index=True, nullable=False, doc='Name of the project, (could also be a proposal ID). ')
+
+    target = sa.Column(sa.Text, index=True, nullable=False, doc='Name of the target object or field id. ')
+
+    gallat = sa.Column(sa.Double, index=True, doc="Galactic latitude of the target. ")
+
+    gallon = sa.Column(sa.Double, index=False, doc="Galactic longitude of the target. ")
+
+    ecllat = sa.Column(sa.Double, index=True, doc="Ecliptic latitude of the target. ")
+
+    ecllon = sa.Column(sa.Double, index=False, doc="Ecliptic longitude of the target. ")
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the exposure object.
+        Can give the filename of the exposure
+        as the single positional argument.
+
+        Otherwise, give any arguments that are
+        columns of the Exposure table.
+
+        If the filename is given, it will parse
+        the instrument name from the filename.
+        The header will be read out from the FITS file.
+        """
+        super().__init__(**kwargs)  # put keywords into columns
+
+        self._data = None
+        self.instrument_object = None
+
+        if len(args) == 1 and isinstance(args[0], str):
+            self.filename = args[0]
+
+        if self.instrument is None:
+            self.parse_instrument_name()
+
+        # find additional column values from the header
+        self.read_header()
+
+        self.num_ccds = self.get_instrument_object().get_num_ccds()
+
+        # override the header values with any given keyword arguments?
+        for key, value in kwargs.items():
+            if key in self.__dict__:
+                setattr(self, key, value)
+
+        if self.ra is not None and self.dec is not None:
+            self.calculate_coordinates()
+
+    @orm.reconstructor
+    def init_on_load(self):
+        self._data = None
+        self.instrument_object = None
+
+    def parse_instrument_name(self):
+        if self.filename is None:
+            return
+
+        for name, class_ in inspect.getmembers(models.instruments, inspect.isclass):
```

Comment:
This is making an assumption about filenames that I don't think we can make -- specifically that there's going to be a unique instrument name in the filename that we can use as a key into an instrument table.  I can envision two failure modes: first, the filename being too generic and not having an instrument name.  Second, the instrument name being a string that shows up in other instruments' filenames.

Filenames of FITS files are not part of a FITS standard or anything like that, so writing code that makes assumptions about those filenames is hazardous.

We need more sophisticated instrument detection.  I would argue for two different things.  First, the ability to just specify it manually : pass in the instrument (name, or database id if we go with a database table of instruments) when instantiating an Exposure.  Second, as a fallback, have a generic "is this exposure from my instrument" method that each instrument class has to implement.  The latter will be slow, as each instrument will go and poke at the file to decide if that file belongs to it, which is why I also want the option to manually specify the instrument.